src/util/amount.rs

```diff
@@ -0,0 +1,921 @@
+    /// Input string was too large.
+    InputTooLarge,
+    /// Invalid character in input.
+    InvalidCharacter(char),
+    /// The denomination was unknown.
+    UnknownDenomination(String),
+}
+
+impl fmt::Display for ParseAmountError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let desc = ::std::error::Error::description(self);
+        match *self {
+            ParseAmountError::InvalidCharacter(c) => write!(f, "{}: {}", desc, c),
+            ParseAmountError::UnknownDenomination(ref d) => write!(f, "{}: {}", desc, d),
+            _ => f.write_str(desc),
+        }
+    }
+}
+
+impl error::Error for ParseAmountError {
+    fn cause(&self) -> Option<&error::Error> {
+        None
+    }
+
+    fn description(&self) -> &'static str {
+        match *self {
+            ParseAmountError::TooBig => "amount is too big",
+            ParseAmountError::TooPrecise => "amount has a too high precision",
+            ParseAmountError::InvalidFormat => "invalid number format",
+            ParseAmountError::InputTooLarge => "input string was too large",
+            ParseAmountError::InvalidCharacter(_) => "invalid character in input",
+            ParseAmountError::UnknownDenomination(_) => "unknown denomination",
+        }
+    }
+}
+
+// The inner type used to represent amounts.
+// This is done to more easily change the underlying type in case this is
+// desired in the future.
+type Inner = i64;
+
+/// Amount
+///
+/// The [Amount] type can be used to express Bitcoin amounts that supports
+/// arithmetic and convertion to various denominations.
+///
+///
+/// Warning!
+///
+/// This type implements several arithmetic operations from [std::ops].
+/// To prevent errors due to overflow or underflow when using these operations,
+/// it is advised to instead use the checked arithmetic methods whose names
+/// start with `checked_`.  The operations from [std::ops] that [Amount]
+/// implements will panic when overflow or underflow occurs.  Also note that
+/// since the internal representation of amounts is unsigned, subtracting below
+/// zero is considered an underflow and will cause a panic if you're not using
+/// the checked arithmetic methods.
+///
+#[derive(Copy, Clone, Hash)]
+pub struct Amount(Inner);
+// The Inner amount represents the number of satoshis.
+
+impl Amount {
+    /// The zero amount.
+    pub const ZERO: Amount = Amount(0);
+    /// Exactly one satoshi.
+    pub const ONE_SAT: Amount = Amount(1);
+    /// Exactly one bitcoin.
+    pub const ONE_BTC: Amount = Amount(100_000_000);
+
+    /// Create a new [Amount] using [amount] as the Inner type.
+    fn from_inner(amount: Inner) -> Amount {
+        Amount(amount)
+    }
+
+    /// Create an [Amount] with satoshi precision and the given number of satoshis.
+    pub fn from_sat(satoshi: i64) -> Amount {
+        Amount::from_inner(satoshi)
+    }
+
+    /// Get the number of satoshis in this [Amount].
+    pub fn as_sat(self) -> i64 {
```

Comment:
Can we have an `as_usat` or something that returns a u64? I see in @jonasnick's code using this that there are zillions of casts to u64.

Reply:
One that panics on negative or one that `Result`s? I'd personally do `try_as_usat() -> Option<u64>` and `as_usat() -> u64`.

Reply:
:( in my usage negative amounts never ever exist. It's really unergonomic to have these extra error paths and also have all my amounts potentially being negative. It also prevents computing fees using `checked_sub` to catch unbalanced transactions as a special case.